### PR TITLE
Warm instances fixes

### PR DESCRIPTION
**Pull request by camilbancioiu**
This PR brings all the fixes for warm instances. Includes older fixes, already reviewed, but the previous default branch of this repo was deleted and replaced with the current `master` branch, which does not contain those older fixes.

Also, this PR adds two conditions which clean warm instances that failed with `ExecutionFailed`.

### CODE REVIEW

File: arwen/helpers.go [hunk 1]
```diff
@@ -141,6 +143,18 @@ type nilInterfaceChecker interface {
 
 // GetVMHost returns the vm Context from the vm context map
 func GetVMHost(vmHostPtr unsafe.Pointer) VMHost {
+	// skipStackLevels := 2
```
Comment (sasurobert) [inline arwen/helpers.go:146]: delete commented code.

File: arwen/hosttest/execution_gas_test.go [hunk 2]
```diff
@@ -958,7 +958,7 @@ func TestGasUsed_AsyncCall_Recursive(t *testing.T) {
 	t.Skip()
 
 	testConfig := &contracts.AsyncCallRecursiveTestConfig{
-		AsyncCallBaseTestConfig: *&asyncBaseTestConfig,
+		AsyncCallBaseTestConfig: asyncBaseTestConfig,
```
Comment (iulianpascalau) [inline arwen/hosttest/execution_gas_test.go:961]: :+1:

File: mock/context/instanceBuilderMock.go [hunk 3]
```diff
@@ -23,6 +23,8 @@ func NewInstanceBuilderMock(world *worldmock.MockWorld) *InstanceBuilderMock {
 	}
 }
 
+func (instance *InstanceBuilderMock) Id() string { return "" }
```
Comment (iulianpascalau) [inline mock/context/instanceBuilderMock.go:26]: missing comment:
```go
// Id -
```

File: mock/context/instanceBuilderRecorderMock.go [hunk 4]
```diff
@@ -29,6 +29,8 @@ func (builder *InstanceBuilderRecorderMock) NewInstanceWithOptions(
 	return instance, err
 }
 
+func (instance *InstanceBuilderRecorderMock) Id() string { return "" }
```
Comment (iulianpascalau) [inline mock/context/instanceBuilderRecorderMock.go:32]: also here, missing mock comment

File: mock/context/instanceMock.go [hunk 5]
```diff
@@ -36,6 +36,8 @@ func NewInstanceMock(code []byte) *InstanceMock {
 	}
 }
 
+func (instance *InstanceMock) Id() string { return "" }
```
Comment (iulianpascalau) [inline mock/context/instanceMock.go:39]: here also
Also, might consider splitting into multiline

File: mock/world/worldAccountMap.go [hunk 6]
```diff
@@ -69,6 +70,10 @@ func (am AccountMap) CreateSmartContractAccountWithCodeHash(owner []byte, addres
 
 // PutAccount inserts account based on address.
 func (am AccountMap) PutAccount(account *Account) {
+	if account.Code != nil && account.CodeHash == nil {
+		hash, _ := hashing.NewHasher().Sha256(account.Code)
```
Comment (iulianpascalau) [inline mock/world/worldAccountMap.go:74]: slightly improve the code by making `hashing.NewHasher()` an unexported package variable

File: test/contracts/bad-empty/elrond.json [hunk 7]
```diff
@@ -0,0 +1,6 @@
+{
+    "language": "clang",
+    "source_files": [
+        "bad-empty.c"
+    ]
+}
```
Comment (iulianpascalau) [inline test/contracts/bad-empty/elrond.json:6]: add empty line?

File: test/contracts/breakpoint/output/breakpoint.imports.json [hunk 8]
```diff
@@ -0,0 +1,5 @@
+[
+    "int64getArgument",
+    "signalError",
+    "int64finish"
+]
```
Comment (iulianpascalau) [inline test/contracts/breakpoint/output/breakpoint.imports.json:5]: add empty line?

File: testcommon/instanceSmartContractCallerTest.go [hunk 9]
```diff
@@ -107,6 +107,10 @@ func (callerTest *InstancesTestTemplate) AndAssertResultsWithoutReset(assertResu
 	runTestWithInstances(callerTest, false)
 }
 
+func (callerTest *InstancesTestTemplate) GetVMHost() arwen.VMHost {
```
Comment (iulianpascalau) [inline testcommon/instanceSmartContractCallerTest.go:110]: missing comment

File: wasmer/instance.go [hunk 10]
```diff
@@ -334,20 +357,43 @@ func (instance *Instance) GetInstanceCtxMemory() MemoryHandler {
 	return instance.InstanceCtx.Memory()
 }
 
+func (instance *Instance) Id() string {
```
Comment (iulianpascalau) [inline wasmer/instance.go:360]: missing comment

**Review by andrei-marinica — CHANGES_REQUESTED** [review]

File: arwen/helpers.go (hunk 1 above)
Comment (camilbancioiu) [reply arwen/helpers.go:146]: Done.

File: arwen/common.go [hunk 11]
```diff
@@ -8,6 +8,7 @@ import (
 // ArwenVersion returns the current arwen version
 const ArwenVersion = "v1.4"
 
+// Size in bytes of a WASM Linear Memory Page
```
[the remark this answers is not in the document]
Comment (camilbancioiu) [reply arwen/common.go:11]: Done.

File: arwen/helpers.go (hunk 1 above)
[the remark this answers is not in the document]
Comment (camilbancioiu) [reply arwen/helpers.go:146]: Done.

File: arwen/host/arwen.go [hunk 12]
```diff
@@ -355,7 +355,7 @@ func (host *vmHost) RunSmartContractCreate(input *vmcommon.ContractCreateInput)
 				log.Error("VM execution panicked", "error", r, "stack", "\n"+string(debug.Stack()))
 				err = arwen.ErrExecutionPanicked
 			}
-			host.Runtime().CleanInstance()
+
```
[the remark this answers is not in the document]
Comment (camilbancioiu) [reply arwen/host/arwen.go:358]: Fixed and improved overall.

File: arwen/hosttest/execution_benchmark_test.go [hunk 13]
```diff
@@ -16,26 +19,56 @@ import (
...
 	worldmock "github.com/ElrondNetwork/wasm-vm-v1_4/mock/world"
 	"github.com/ElrondNetwork/wasm-vm-v1_4/testcommon"
+	"github.com/stretchr/testify/assert"
 	"github.com/stretchr/testify/require"
 )
 
-var owner = []byte("owner")
-var receiver = []byte("receiver")
-var scAddress = []byte("erc20")
+type Address = []byte
```
[the remark this answers is not in the document]
Comment (camilbancioiu) [reply arwen/hosttest/execution_benchmark_test.go:26]: Added.

File: mock/context/instanceBuilderMock.go (hunk 3 above)
Comment (camilbancioiu) [reply mock/context/instanceBuilderMock.go:26]: Added.

File: mock/context/instanceMock.go (hunk 5 above)
Comment (camilbancioiu) [reply mock/context/instanceMock.go:39]: Done.

File: mock/world/worldAccountMap.go (hunk 6 above)
Comment (camilbancioiu) [reply mock/world/worldAccountMap.go:74]: Done.

File: arwen/helpers.go (hunk 1 above)
[the remark this answers is not in the document]
Comment (camilbancioiu) [reply arwen/helpers.go:146]: Removed.

File: wasmer/instance.go (hunk 10 above)
Comment (camilbancioiu) [reply wasmer/instance.go:360]: Added.

File: testcommon/instanceSmartContractCallerTest.go (hunk 9 above)
Comment (camilbancioiu) [reply testcommon/instanceSmartContractCallerTest.go:110]: Added.

**Review by andrei-marinica — APPROVED** [review]

File: arwen/host/execution.go [hunk 14]
```diff
@@ -27,9 +27,17 @@ func (host *vmHost) doRunSmartContractCreate(input *vmcommon.ContractCreateInput
 
 	_, blockchain, metering, output, runtime, storage := host.GetContexts()
 
+	var vmOutput *vmcommon.VMOutput
+	defer func() {
+		if vmOutput.ReturnCode == vmcommon.ExecutionFailed {
```
[the remark this answers is not in the document]
Comment (iulianpascalau) [reply arwen/host/execution.go:32]: Right, we have discussed here that we should have either:
```go
if vmOutput != nil && vmOutput.ReturnCode == vmcommon.ExecutionFailed {
```
or 
```go
if vmOutput == nil || vmOutput.ReturnCode == vmcommon.ExecutionFailed {
```

File: arwen/host/execution.go [hunk 15]
```diff
@@ -113,14 +118,22 @@ func (host *vmHost) doRunSmartContractUpgrade(input *vmcommon.ContractCallInput)
 
 	_, _, metering, output, runtime, storage := host.GetContexts()
 
+	var vmOutput *vmcommon.VMOutput
+	defer func() {
+		if vmOutput.ReturnCode == vmcommon.ExecutionFailed {
```
Comment (iulianpascalau) [inline arwen/host/execution.go:123]: also here, same suggestion as above

File: arwen/host/execution.go [hunk 16]
```diff
@@ -159,6 +173,13 @@ func (host *vmHost) doRunSmartContractCall(input *vmcommon.ContractCallInput) (v
 
 	_, _, metering, output, runtime, storage := host.GetContexts()
 
+	var vmOutput *vmcommon.VMOutput
+	defer func() {
+		if vmOutput == nil || vmOutput.ReturnCode == vmcommon.ExecutionFailed {
```
Comment (iulianpascalau) [inline arwen/host/execution.go:178]: :+1:

**Review by iulianpascalau — CHANGES_REQUESTED** [review]

File: arwen/host/execution.go (hunk 14 above)
[the remark this answers is not in the document]
Comment (camilbancioiu) [reply arwen/host/execution.go]: Added forgotten `nil` check to the condition...

File: arwen/host/execution.go (hunk 15 above)
Comment (camilbancioiu) [reply arwen/host/execution.go]: Added now...

**Review by iulianpascalau — APPROVED** [review]

**Review by andrei-marinica — APPROVED** [review]

**Review by sasurobert — APPROVED** [review]

**Review by valentin-lup — APPROVED** [review]
preliminary tests with importdb passed !

**Pull request merged**